text/0000-typed-context-injection.md

```diff
@@ -0,0 +1,833 @@
+}
+
+pub struct SoundVolumeOverrides { ... }
+
+impl SoundVolumeOverrides {
+    pub fn get_volume_override(path: &str) -> f32 { ... }
+}
+
+// Ideally, we could just write:
+cx.play_sound_at_path("cbat.ogg");
+
+// But, unfortunately, we must write:
+audio_buffer.play_sound_at_path(&mut virtual_fs, &mut audio_cache, &mut sound_volume_overrides, "cbat.ogg");
+```
+
+Yuck!
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Enter typed context injection. 
+
+Context injection is accomplished with two new standard library items: `Cx` and `AnyCx`. `Cx` is a variadic type parameterized by the references comprising a function's context. For example, to write a function taking `&mut System1` and `&System2`, you can just write:
+
+```rust
+fn consumer(cx: Cx<&mut System1, &System2>) { ... }
+```
+
+From there, you can fetch your components by either coercing the context into its comprising elements (which we'll henceforth refer to as *components*) or by using the `.extract::<T>()` and `.extract_mut::<T>()` utility methods, which perform the coercion internally but expose it in an occasionally-more-convenient method form.
+
+```rust
+fn consumer(cx: Cx<&mut System1, &System2>) {
+    let system_1: &mut System1 = cx;
+    let system_2: &System2 = cx;
+
+    // ...which is equivalent to:
+    let system_1 = cx.extract_mut::<System1>();
+    let system_2 = cx.extract::<System2>();
+}
+```
+
+To provide a `Cx` to a method, you can either create it with the `Cx::new` constructor or coerce a superset context into the target subset of itself.
+
+```rust
+fn caller_1(cx: Cx<&mut System1, &mut System2, &System3>) {
+    // `cx` has a mutable reference to `System1` and (at least) an immutable reference to `System2`
+    // so this is fine. The unnecessary `&System3` reference is completely ignored.
+    consumer(cx);
+}
+
+fn caller_2() {
+    ...
+
+    // We can give the context in the order it shows up in the target type...
+    consumer(Cx::new((&mut system_1, &system_2)));
+
+    // ...but we could also provide the context in any other order and let coercion
+    // turn e.g. a `Cx<&mut System2, &mut System1>` into a `Cx<&mut System1, &System2>`.
+    consumer(Cx::new((&mut system_2, &mut system_1)));
+
+    // Heck, we could even pass the function additional completely unrelated components and coercion
+    // would still have our back!
+    //
+    // Coerces `Cx<&mut SomeOtherSystem, &mut System2, &mut System1>` into
+    // `Cx<&mut System1, &System2>`.
+    consumer(Cx::new((&mut some_other_system, &mut system_2, &mut system_1)));
+
+    // And if that flexibility wasn't enough, we can combine several contexts together to
+    // construct a full context like so:
+    let inherited_cx = Cx::new((&mut some_other_system, &mut system_2));
+
+    // Coerces `Cx<Cx<&mut SomeOtherSystem, &mut System2>, &mut System1>` into
+    // `Cx<&mut System1, &System2>`.
+    consumer(Cx::new((inherited_cx, &mut system_1)));
```

Comment:
If you write out the type:

```rust
type MyType<'a> = Cx<Cx<&'a System1, &'a System2>, &'a System1>;
```

It would indeed be deduplicated to just:

```rust
type MyType<'a> = Cx<&'a System1, &'a System2>;
```

However, if you wrote:

```rust
let foo = Cx::new((&system_1, &system_2));
let bar = Cx::new((foo, &system_1));
```

The types would resolve as follows:

```
let foo = Cx::new((&system_1, &system_2));  // `foo` has the type `CxRaw<(&System1, &System2)>`.
let bar = Cx::new((foo, &system_1));  // `bar` has the type `CxRaw<(CxRaw<(&System1, &System2)>, &System1)>`.
```

Reply:
`Cx` itself is a type. Why can't I store a context inside a context?

Reply:
> `Cx` itself is a type. Why can't I store a context inside a context?

You absolutely can store a context in a context—it's just that, when constructing the actual object, it will not be flattened until it needs to.

That... was a bit confusing so let me clarify.

Deduplication and flattening happen at the `Cx` level and the coercion level, not the `CxRaw` level. `Cx` is a type alias for `CxRaw` which eagerly deduplicates the component list to the best of its ability and only does so once. If that list contains generics, generics will be deduplicated based off their known equalities, not the potential equalities brought on by their substitution by the user. `CxRaw` is the thing you actually construct when you call `Cx::new((..., ...))` and its type is exactly what gets inferred by its constructor. When I say, "it will not be flattened until it needs to," I mean that `CxRaw<(CxRaw<&System1, &System2>, &System3)>` will only be flattened to a `CxRaw<(&System1, &System2, &System3)>` during coercion. This isn't something users tend to worry about, however, since this coercion is pretty much invisible.

You might want to reread the section about generics since that explains these rules in much more detail.